scene/main/viewport.cpp

```diff
@@ -2442,6 +2442,94 @@ void Viewport::_gui_remove_control(Control *p_control) {
 	}
 }
 
+void Viewport::canvas_item_top_level_changed() {
+	_gui_update_mouse_over();
+}
+
+void Viewport::_gui_update_mouse_over() {
+	if (gui.mouse_over == nullptr || gui.mouse_over_hierarchy.is_empty()) {
+		return;
+	}
+
+	// Rebuild the mouse over hierarchy.
+	LocalVector<Control *> new_mouse_over_hierarchy;
+	LocalVector<Control *> needs_enter;
+	LocalVector<int> needs_exit;
```

Comment:
We may want to avoid allocations in such a hot function. @Sauermann has an idea on maybe using just references to nodes at the ends of the chain. Maybe the other vector needs something else. This is just an idea about optimization, not meant as a blocker for this PR in any case.

Reply:
I'm not sure this function is very hot. It is only called when `set_top_level()` or `set_mouse_filter()` is called, and only executed if the mouse is over something.
If we want I can add a guard before the allocations that is similar to the loop below it but smaller.

Reply:
Also the `mouse_over_hierarchy` contains a list of all nodes that have gotten `mouse_entered` but still need a `mouse_exited`. If we change this to only store the ends, we wouldn't be able to update properly when the mouse filter ignore is set.